File: app/cpp/src/examples/faspex5.cpp
#include <iostream>
#include <stdexcept>
#include <string>

#include "utils/configuration.hpp"
#include "utils/rest.hpp"
#include "utils/transfer_client.hpp"

const std::string F5_API_PATH_V5 = "/api/v5";
const std::string F5_API_PATH_TOKEN = "/auth/token";
const std::string package_name = "sample package";
const int transfer_sessions = 1;  // Typically, 1

int main(const int argc, const char* const argv[]) {
    utils::Configuration config(argc, argv);
    utils::TransferClient transfer_client(config);
    try {
        // Create the REST client for the Faspex 5 API
        std::string faspex_url = config.param_str({"faspex5", "url"}) + F5_API_PATH_V5;
        utils::Rest f5_api(faspex_url);
        f5_api.set_verify(config.param_bool({"faspex5", "verify"}, true));

        // Set up bearer token authentication
        f5_api.set_auth_bearer(  //
            {{"token_url", config.param_str({"faspex5", "url"}) + F5_API_PATH_TOKEN},
             {"key_pem_path", config.param_str({"faspex5", "private_key"})},
             {"client_id", config.param_str({"faspex5", "client_id"})},
             {"client_secret", config.param_str({"faspex5", "client_secret"})},
             {"iss", config.param_str({"faspex5", "client_id"})},
             {"aud", config.param_str({"faspex5", "client_id"})},
             {"sub", "user:" + config.param_str({"faspex5", "username"})}});

        f5_api.set_default_scope("");

        // Create a new package with Faspex 5 API
        LOG(info) << "Creating package: " << package_name;
        json::object package_info = f5_api.create(
            "packages",
            {{"title", package_name},
             {"recipients", json::array{
                                json::object{
                                    {"name", config.param_str({"faspex5", "username"})}}}}}).as_object();
        LOG(debug) << package_info;

        // Build payload to specify files to send
        json::object upload_request = json::object{};
        config.add_sources(upload_request, "paths");

        LOG(info) << "Getting transfer spec";
        std::ostringstream endpoint;
        endpoint << "packages/" << package_info.at("id").as_string().c_str() << "/transfer_spec/upload";
        LOG(info) << ">>>>>>" << endpoint.str();
        json::object t_spec = f5_api.create(endpoint.str(), upload_request, {{"transfer_type", "connect"}}).as_object();

        // Optional: Multi session
        if (transfer_sessions != 1) {
            t_spec.at("multi_session") = transfer_sessions;
            t_spec.at("multi_session_threshold") = 500000;
        }

        // Add file list in transfer spec
        config.add_sources(t_spec, "paths");

        // Not used in transfer sdk
        t_spec.erase("authentication");

        // Finally, send files to package folder on server
        transfer_client.transfer_start_and_wait(t_spec);
        return 0;
    } catch (const std::exception& e) {
        std::clog << "Exception: " << e.what() << std::endl;
        return 1;
    }
}

File: app/go/src/examples/faspex5.go
package main

import (
	"aspera_examples/src/utils"
	"fmt"
)

const (
	F5APIPathV5      = "/api/v5"
	F5APIPathToken   = "/auth/token"
	packageName      = "sample package"
	transferSessions = 1
)

func main() {
	config, err := utils.NewConfiguration()
	if err != nil {
		config.Log.Fatalf("Error loading configuration: %v", err)
	}
	transferClient := utils.NewTransferClient(config)
	defer transferClient.Shutdown()

	// Generate a bearer token for each script invocation
	f5API := utils.NewRest(fmt.Sprintf("%s%s", config.ParamStr("faspex5", "url"), F5APIPathV5))
	f5API.SetVerify(config.ParamBool("faspex5", "verify", true))
	f5API.SetBearer(map[string]string{
		"token_url":     fmt.Sprintf("%s%s", config.ParamStr("faspex5", "url"), F5APIPathToken),
		"key_pem_path":  config.ParamStr("faspex5", "private_key"),
		"client_id":     config.ParamStr("faspex5", "client_id"),
		"client_secret": config.ParamStr("faspex5", "client_secret"),
		"iss":           config.ParamStr("faspex5", "client_id"),
		"aud":           config.ParamStr("faspex5", "client_id"),
		"sub":           fmt.Sprintf("user:%s", config.ParamStr("faspex5", "username")),
	})
	f5API.SetDefaultScope("")
	// Create a new package with Faspex 5 API
	config.Log.Debugf("Creating package: %s", packageName)
	packageResp, err := f5API.Create("packages", map[string]interface{}{
		"title":      packageName,
		"recipients": []map[string]string{{"name": config.ParamStr("faspex5", "username")}},
	})
	if err != nil {
		config.Log.Fatalf("Failed to create package: %v", err)
	}
	config.Log.Debugf("Package info: %+v", packageResp)

	// Build payload to specify files to send
	filesToSend := map[string]interface{}{"paths": []map[string]string{}}
	config.AddSources(filesToSend, "paths")

	config.Log.Debugf("Getting transfer spec")
	tSpec, err := f5API.Create(fmt.Sprintf("packages/%v/transfer_spec/upload?transfer_type=connect", packageResp["id"]), filesToSend)
	if err != nil {
		config.Log.Fatalf("Failed to get transfer spec: %v", err)
	}

	// Optional: multi-session
	if transferSessions != 1 {
		tSpec["multi_session"] = transferSessions
		tSpec["multi_session_threshold"] = 500000
	}

	// Add file list in transfer spec
	tSpec["paths"] = []map[string]string{}
	config.AddSources(tSpec, "paths")

	// Remove authentication (not used in transfer sdk)
	delete(tSpec, "authentication")

	// Finally send files to package folder on server
	transferClient.StartTransferAndWait(tSpec)
}
